### DISCUSSION TITLE
How exactly does the wasm audio worklet API works?

### DISCUSSION THREAD

**Discussion by msqr1** [emscripten-core/emscripten#21714] [Q&A] [answered]
On the docs, it says that the audio worklet runs in a wasm worker, what does it mean? Doesn't creating the audio worklet create another thread already?
It also says that the user will not immplement the ```ScriptProcessorNode``` by themselves, but isn't the audio worklet uses the audioworklet and not the deprecated ```ScriptProcessorNode```

**Comment by sbc100** [comment]
@juj

**Comment by msqr1** [comment]
@juj you implemented this, right?

**Comment by sbc100** [comment]
I assume you read the documentation on this: https://emscripten.org/docs/api_reference/wasm_audio_worklets.html ?

**Comment by msqr1** [comment]
Yeah

**Comment by sbc100** [comment]
In terms of the relationship between wasm workers and audio worklet, IIUC and audio worklet in emscripten is implemented as a wasm worker.  The wasm worker runs *on* the audio worklet thread, so there is exactly one worker created when you create an audio worklet.

I think the docs mention `ScriptProcessorNode` only to say that you don't need to worry about implementing any of the stuff at that level when using this API.  emscripten takes care of the internals.  And yes, we don't use `ScriptProcessorNode` internally.

**Comment by msqr1** [comment]
Why does the wasm worker run on top of the audio worklet? Didn't we instantiate the wasm module in audioworkletglobalscope, where is the part that a wasm worker is required?

One more thing, could we allow the user to specify block size other than the default 128? I think this can be easily implemented.

**Reply by sbc100** [reply, marked as answer]
A Wasm Worker is just nothing really more than some code to host/run run the wasm module.   Its the minimal thing that is required for the compiled code to be able to communicate with other workers via shared memory.

Perhaps open a new issue to discuss the block size (I have vague memories of it being discussed elsewhere).

**Reply by msqr1** [reply]
Ok
